src/micro.c

```diff
@@ -0,0 +1,680 @@
+    *ptr = NATS_CALLOC(1, sizeof(microService));
+    if (*ptr == NULL)
+        return micro_ErrorOutOfMemory;
+
+    natsConn_retain(nc);
+    (*ptr)->refs = 1;
+    (*ptr)->nc = nc;
+    (*ptr)->started = nats_Now() * 1000000;
+    return NULL;
+}
+
+microService *
+micro_retain_service(microService *m)
+{
+    if (m == NULL)
+        return NULL;
+
+    micro_lock_service(m);
+
+    m->refs++;
+
+    micro_unlock_service(m);
+
+    return m;
+}
+
+void micro_release_service(microService *m)
+{
+    int refs = 0;
+
+    if (m == NULL)
+        return;
+
+    micro_lock_service(m);
+
+    refs = --(m->refs);
+
+    micro_unlock_service(m);
+
+    if (refs == 0)
+        free_service(m);
+}
+
+static void free_service(microService *m)
+{
+    microGroup *next = NULL;
+
+    if (m == NULL)
+        return;
+
+    // destroy all groups.
+    if (m->groups != NULL)
+    {
+        microGroup *g = m->groups;
+        while (g != NULL)
+        {
+            next = g->next;
+            NATS_FREE(g);
+            g = next;
+        }
+    }
+
+    micro_free_cloned_service_config(m->cfg);
+    natsConn_release(m->nc);
+    natsMutex_Destroy(m->service_mu);
+    NATS_FREE(m);
+}
+
+static inline microError *new_service_config(microServiceConfig **ptr)
+{
+    *ptr = NATS_CALLOC(1, sizeof(microServiceConfig));
+    return (*ptr == NULL) ? micro_ErrorOutOfMemory : NULL;
+}
+
+microError *
+micro_clone_service_config(microServiceConfig **out, microServiceConfig *cfg)
+{
+    microError *err = NULL;
+    microServiceConfig *new_cfg = NULL;
+
+    if (out == NULL || cfg == NULL)
+        return micro_ErrorInvalidArg;
+
+    err = new_service_config(&new_cfg);
+    if (err == NULL)
+    {
+        memcpy(new_cfg, cfg, sizeof(microServiceConfig));
+    }
+    // the strings are declared const for the public, but in a clone these need
+    // to be duplicated.
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Name, cfg->Name));
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Version, cfg->Version));
+    MICRO_CALL(err, micro_strdup((char **)&new_cfg->Description, cfg->Description));
+    MICRO_CALL(err, micro_clone_endpoint_config(&new_cfg->Endpoint, cfg->Endpoint));
+    if (err != NULL)
+    {
+        micro_free_cloned_service_config(new_cfg);
+        return err;
+    }
+
+    *out = new_cfg;
+    return NULL;
+}
+
+void micro_free_cloned_service_config(microServiceConfig *cfg)
+{
+    if (cfg == NULL)
+        return;
+
+    // the strings are declared const for the public, but in a clone these need
+    // to be freed.
+    NATS_FREE((char *)cfg->Name);
+    NATS_FREE((char *)cfg->Version);
+    NATS_FREE((char *)cfg->Description);
+    micro_free_cloned_endpoint_config(cfg->Endpoint);
+    NATS_FREE(cfg);
+}
+
+static void
+on_connection_closed(natsConnection *nc, void *closure)
+{
+    microService *m = (microService *)closure;
+    if (m == NULL)
+        return;
+
+    microService_Stop(m);
+
+    if (m->prev_on_connection_closed != NULL)
```

Comment:
Name seems to indicate that this is a list, but am I correct that `on_connection_closed` is invoked because we have set our own callback (that one) and now want to invoke the user's callback? Again, the name indicates that it is a list, but I don't see where we would do the chaining, and if we do, then is there a thread-safety issue?
If my understanding that there is no list but just the micro one and then the user one, then I thing `user_on_connection_closed_cb` or something like that would be more explicit.

Reply:
It is not a list, and that is actually a bit of a problem - they need to be unwound in the _exact_ reverse order by the callers. I intended to rework it into a list, but perhaps as a separate PR? At the moment it is only impacting the micro-sequence example.

Reply:
[See #656](https://github.com/nats-io/nats.c/pull/656) addresses this. Resolving this (and related) comments, we can continue the conversation there.